### PR TITLE
Expose license status in Info

### PR DESCRIPTION
**Pull request by dhiltgen**
**- What I did**

This wires up a new field in the `Info` payload that exposes the license.
For moby this is hardcoded to always report a community edition string.
Downstream enterprise dockerd will have additional licensing logic wired
into this function to report details about the current license status.

**- How I did it**

**- How to verify it**

Simple unit test included.  Until the Info structure shows up in the CLI, this can be seen via curl
```
% curl -s --unix-socket /var/run/docker.sock http:/info  | jq ".License"
"Community Edition Engine"
```

**- Description for the changelog**

Exposes license information in Info output

**- A picture of a cute animal (not mandatory but encouraged)**

screen_2018-08-08-15-26-47

### CODE REVIEW

**Comment by thaJeztah** [comment]
The Docker Enterprise platform consists of multiple components (Engine, UCP, DTR, possibly other components). I'm wondering if this information should be included in the `Components` information of the `/version` endpoint https://docs.docker.com/engine/api/v1.37/#operation/SystemVersion

**Comment by thaJeztah** [comment]
Also; "license" as a term could be ambiguous here (i.e., the software is Apache Licensed)

**Comment by dhiltgen** [comment]
> The Docker Enterprise platform consists of multiple components (Engine, UCP, DTR, possibly other components). I'm wondering if this information should be included in the Components information of the /version endpoint https://docs.docker.com/engine/api/v1.37/#operation/SystemVersion

Hmm... I don't really think of the license as a versioned component, so it feels like that's not quite the right place to put it.  UCP does augment the components API and include information about itself, but doesn't include licensing details there.  Our goal with the upcoming release is to have the same license applied to the engine and UCP/DTR, so whatever we expose in the API/CLI will by virtue of that connection be shared across the cluster.


> Also; "license" as a term could be ambiguous here (i.e., the software is Apache Licensed)

I'm open to suggestions on a better term.  Perhaps "ProductLicense"?

File: daemon/licensing.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package daemon // import "github.com/docker/docker/daemon"
+
+import (
+	"github.com/docker/docker/api/types"
+)
+
+func (daemon *Daemon) fillLicense(v *types.Info) {
+	v.ProductLicense = "Community Engine"
```
Comment (thaJeztah) [inline daemon/licensing.go]: Question; will we actually use this string for the Community Engine, or is this just to "not return an empty string"? If we don't need this string, we could consider leaving this empty

File: api/types/types.go [hunk 2]
```diff
@@ -204,6 +204,7 @@ type Info struct {
 	RuncCommit         Commit
 	InitCommit         Commit
 	SecurityOptions    []string
+	ProductLicense     string
```
Comment (thaJeztah) [inline api/types/types.go]: For Enterprise Engines; is a single string all the information we need? (e.g. expiry date of the license, type of license and such?)

**Comment by thaJeztah** [comment]
> I'm open to suggestions on a better term. Perhaps "ProductLicense"?

Product license SGTM

> UCP does augment the components API and include information about itself, but doesn't include licensing details there. Our goal with the upcoming release is to have the same license applied to the engine and UCP/DTR, so whatever we expose in the API/CLI will by virtue of that connection be shared across the cluster.

That's was the bit I was curious about, e.g. if there's plans to have individual licenses for components (in which case license for the "engine" could be included in the "engine" component information, and licenses for other components in their respective part of the response). I'm ok with the current implementation if none of those will be the case.

One concern I had was that the `/info` endpoint is known to not be really _lightweight_. There's a _lot_ of information that's collected by this endpoint, and some of that information is not "cheap"; that's not an issue if this information is only requested occasionally, but thought I'd mention.

Final consideration; do we need this information exposed through the `/nodes` endpoint as well (for multi-node setups?).

> may need an update to the swagger.yml and API changelog

Yes; this should be included in the swagger file; https://github.com/moby/moby/blob/bd06a5ea4df919ff323510fba10801b5673a3af6/api/swagger.yaml?utf8=✓#L3463-L3465, and the API changelog; https://github.com/moby/moby/blob/master/docs/api/version-history.md#v139-api-changes


I'll leave some other comments inline

File: daemon/licensing.go (hunk 1 above)
Comment (dhiltgen) [reply daemon/licensing.go]: Hiding the field on community engines could lead to ambiguity/confusion - I think it's better to be explicit.

File: api/types/types.go (hunk 2 above)
Comment (dhiltgen) [reply api/types/types.go]: Covered by the licensing library - see https://github.com/docker/licensing/blob/master/model/subscriptions.go#L38

I'm working on a change for the enterprise engine to wire it up and hope to have that posted in the next day or two.

(I think we'll likely refine the exact string a bit as we finalize UX but the intention is a human friendly representation of the license)

**Comment by dhiltgen** [comment]
> One concern I had was that the /info endpoint is known to not be really lightweight. There's a lot of information that's collected by this endpoint, and some of that information is not "cheap"; that's not an issue if this information is only requested occasionally, but thought I'd mention.

I think the /info API is a good place to expose this information for this release.  This wont be used by clients to perform license validation, but rather as a means to expose the current license status to the user.

> Final consideration; do we need this information exposed through the /nodes endpoint as well (for multi-node setups?).

In a clustered setting, a single license is applied to the entire cluster and the license entitles the user to a certain number of nodes.   At the moment I'm not aware of any requirements to support divergent licenses across a cluster.

**Comment by thaJeztah** [comment]
LGTM after the swagger-file and API changelog is updated

**Comment by cpuguy83** [comment]
> wdyt on the design of this PR?

It's definitely off-putting. I understand the requirement but feels wrong as implemented.

Spitballing, but if the endpoint had an "extensions" field which was `map[string]interface{}` (or `map[string]string` could work just not as flexible) then Docker clients could pull the license details out of there.

**Comment by dhiltgen** [comment]
> LGTM after the swagger-file and API changelog is updated

Updated.

**Comment by tiborvass** [comment]
@dhiltgen I kinda agree with @cpuguy83, couldn't we set it to an empty string by default and allow it to be set at build-time just like we do for all those variables in hack/make/.go-autogen ? That way CE can set it to CE, EE can set it to EE licenses but it stays empty in moby. Make sure you add `json:",omitempty"`.

@cpuguy83 what do you think?

**Comment by dhiltgen** [comment]
> @dhiltgen I kinda agree with @cpuguy83, couldn't we set it to an empty string by default and allow it to be set at build-time just like we do for all those variables in hack/make/.go-autogen ? That way CE can set it to CE, EE can set it to EE licenses but it stays empty in moby. Make sure you add json:",omitempty".

Sounds reasonable to me.  If folks are :+1: I'll make the change.

**Comment by cpuguy83** [comment]
1. A simple string is likely not enough, though I suppose the entire license can be encoded as a string...
2. Keep in mind this is moby. Docker wants to introduce licensing, the licensing logic should be in the Docker product and moby should be made to be able support this extensibility.
3. I do not think a straight up "license" field satisfies 2.

**Comment by tiborvass** [comment]
I understand your concern, though I believe making the field an empty string in moby is good enough. We already do this for other variables and anyone can reuse them.

LGTM

**Review by tiborvass — APPROVED** [review]

**Pull request merged**